ykrt/src/compile/jitc_yk/trace_builder.rs

```diff
@@ -1,44 +1,127 @@
+    ///  - `trace_name`: The eventual symbol name for the JITted code.
+    ///  - `aot_mod`: The AOT IR module that the trace flows through.
+    ///  - `mtrace`: The mapped trace.
+    fn new(trace_name: String, aot_mod: &'a Module, mtrace: &'a Vec<TracedAOTBlock>) -> Self {
         Self {
             aot_mod,
             mtrace,
-            jit_mod: Module::default(),
+            jit_mod: jit_ir::Module::new(trace_name),
+            local_map: HashMap::new(),
         }
     }
 
-    fn lookup_aot_block(&self, tb: &TracedAOTBlock) -> Option<&aot_ir::Block> {
+    // Given a mapped block, find the AOT block ID, or return `None` if it is unmapped.
+    fn lookup_aot_block(&self, tb: &TracedAOTBlock) -> Option<aot_ir::BlockID> {
         match tb {
             TracedAOTBlock::Mapped { func_name, bb } => {
                 let func_name = func_name.to_str().unwrap(); // safe: func names are valid UTF-8.
-                let func = self.aot_mod.func_by_name(func_name)?;
-                func.block(*bb)
+                let func = self.aot_mod.func_index(func_name)?;
+                Some(aot_ir::BlockID::new(func, *bb))
             }
             TracedAOTBlock::Unmappable { .. } => None,
         }
     }
 
-    fn build(self) -> Result<Module, Box<dyn Error>> {
+    /// Create the prolog of the trace.
+    fn create_trace_header(&mut self, blk: &aot_ir::Block) {
+        // Find trace input variables and emit `LoadArg` instructions for them.
+        let mut last_store = None;
+        let mut trace_input = None;
+        let mut input = Vec::new();
+        for inst in blk.instrs.iter().rev() {
+            if inst.is_control_point(self.aot_mod) {
+                trace_input = Some(inst.get_operand(3));
+            }
+            if inst.is_store() {
+                last_store = Some(inst);
+            }
+            if inst.is_gep() {
+                let op = inst.get_operand(0);
+                // unwrap safe: we know the AOT code was produced by ykllvm.
+                if trace_input.unwrap().to_instr(self.aot_mod) == op.to_instr(self.aot_mod) {
+                    // Found a trace input.
+                    // unwrap safe: we know the AOT code was produced by ykllvm.
+                    let inp = last_store.unwrap().get_operand(0);
+                    input.insert(0, inp.to_instr(self.aot_mod));
+                    let load_arg = jit_ir::Instruction::create_loadarg();
+                    self.local_map.insert(inp.to_instr_id(), self.jit_mod.len());
+                    self.jit_mod.push(load_arg);
+                }
+            }
+        }
+    }
+
+    /// Walk over a traced AOT block, translating the constituent instructions into the JIT module.
+    fn process_block(&mut self, bid: aot_ir::BlockID) {
+        // unwrap safe: can't trace a block not in the AOT module.
+        let blk = self.aot_mod.block(&bid).unwrap();
+
+        // Decide how to translate each AOT instruction based upon it's opcode.
```

Comment:
s/it's/its/ (yes, I am a member of the [Apostrophe Protection Society](https://www.apostrophe.org.uk/) :p)

Reply:
I dunno man, looks like you wanna kill that apostrophe! :laughing:

Reply:
Fixed in c02dd73a02dede5bd706e6567f1532367daf31d9